### PR TITLE
Implement a `repeat_until_end` format.

### PR DESCRIPTION
**Pull request by brendanzab**
This implements a `repeat_until_end` format, along with an `Array` type with no length. We could add more repeat formats later in order to address the needs of #352.

These commits were taken from #310.

### CODE REVIEW

**Review by wezm — APPROVED** [review]
Looks good 👍 I was expecting it to require more code when I read the PR description.

File: doc/reference.md [hunk 1]
```diff
@@ -355,6 +356,23 @@ of the host array types.
...
+The `repeat_until_end` format repeats parsing the given format until the end of
+the current binary stream is reached:
+
+- `repeat_until_end : Format -> Format`
+
+#### Representation of repeat formats
+
+Because the repeat format does not have a predefined length, it is
+[represented](#format-representations) as a dynamically sized
+[array type](#array-types):
```
Comment (mikeday) [inline doc/reference.md:370]: Is it possible to access the length once parsed? Do we need to mention that the repeat format doesn't need to be of fixed size?

File: fathom/src/core/binary.rs [hunk 2]
```diff
@@ -183,7 +184,33 @@ impl<'arena, 'env> Context<'arena, 'env> {
...
+                    elems.push(elem);
+                    current_pos = reader.stream_position()?;
+                }
+                Err(err) => match err.kind() {
+                    io::ErrorKind::UnexpectedEof => {
+                        // FIXME: should this be set to the end of the current stream?
+                        reader.seek(SeekFrom::Start(current_pos))?;
+                        return Ok(Arc::new(Value::ArrayLit(elems)));
+                    }
+                    _ => return Err(err),
```
Comment (mikeday) [inline fathom/src/core/binary.rs:207]: Does this allow trailing random bytes at the end of the stream?

Comment (brendanzab) [reply fathom/src/core/binary.rs:207]: Ahh yeah it does - I was thinking of maybe adding a `repeat_until_end_exact` format.

File: doc/reference.md (hunk 1 above)
Comment (brendanzab) [reply doc/reference.md:370]: Yeah, I think I'll need to get up to adding sizing to this reference, like how I talk about the host representation types.

Are you thinking the byte length? Or the number of elements in the array?

**Comment by brendanzab** [comment]
Sorry for being hasty in merging this @mikeday! 😳

File: doc/reference.md (hunk 1 above)
Comment (mikeday) [reply doc/reference.md:370]: I was thinking of number of elements in the array, as the byte length could be determined by following it with a pos.

**Comment by mikeday** [comment]
No problems with the merge, just curious!

**Pull request merged**